File: Projects/Software Technologies/Exam Preparation 1/C# Skeleton/TodoList/Controllers/TaskController.cs
using System;
using System.Linq;
using System.Web.Mvc;
using TodoList.Models;

namespace TodoList.Controllers
{
	public class TaskController : Controller
	{
	    [HttpGet]
        [Route("")]
        ////This method will render the home view
        public ActionResult Index()
	    {
            using (var db = new TodoListDbContext())
            {
                //this will take all tasks from the databse and put them in a List<Task>.
                var tasks = db.Tasks.ToList();

                //Then we will give them to the Home View (Index.cshtml)
                return View(tasks);
            }
	       
        }

        [HttpGet]
        [Route("create")]
        //This method will only return the view for creating a task.
        public ActionResult Create()
		{
		    return View();//this will return the view in Views/Task/Create.cshtml.
		}

		[HttpPost]
		[Route("create")]
        [ValidateAntiForgeryToken]
        //This method will be used for submitting the info from the form to the database.
		public ActionResult Create(Task task)
		{
            //We check if the task is null and if this condition is true we redirect to Home Page.
            if (task == null)
            {
                return RedirectToAction("Index");
            }

            //We check if the task is with a missing title or comments
            //and if that conditon is true we redirect to Home Page.
            if (string.IsNullOrWhiteSpace(task.Title) 
                || string.IsNullOrWhiteSpace(task.Comments))
            {
                return RedirectToAction("Index");
            }

            //If none of the ifs before returns true we create connection with the database
            // using variable called db.
            using (var db = new TodoListDbContext())
            {
                //With db we can do things with the database like find, add, remove, etc...

                //So we add the task which we took from the create form to the database
                db.Tasks.Add(task);
                //And we save the changes.
                db.SaveChanges();

                //Finally we render to view to Home Page.
                return RedirectToAction("Index");
            }
		    
        }

		[HttpGet]
		[Route("delete/{id}")]
        public ActionResult Delete(int id)
		{
            //With db we create connection with the database.
            //Also with db we can do things with the database like find, add, remove, etc...
            using (var db = new TodoListDbContext())
            {
                //We find the task we want to delete by id.
                //and we save this task in an task variable of type Task.
                var task = db.Tasks.Find(id);

                //We check if the task is null and if this condition is true we redirect to Home Page.
                if (task == null)
                {
                    return RedirectToAction("Index");
                }
                //finally we render the Delete.cshtml view
                //And we put the task in it.
                return View(task);
            }
		    
        }

		[HttpPost]
		[Route("delete/{id}")]
        [ValidateAntiForgeryToken]
		public ActionResult DeleteConfirm(int id)
		{
            //With db we can do things with the database like find, add, remove, etc...
            using (var db = new TodoListDbContext())
            {
                //We find the task we want to delete in the databse by id.
                //and we save this task in an task variable of type Task.
                var task = db.Tasks.Find(id);

                //We check if the task is null and if this condition is true we redirect to Home Page.
                if (task == null)
                {
                    return RedirectToAction("Index");
                }

                //Else if the task is not null we remove it from the databse.
                //We use Remove() and we give this function the task, which we want to remove
                //In our example that is the task stored in the task variable.
                db.Tasks.Remove(task);
                //We save the changes in the databse.
                db.SaveChanges();

                //Finally we redirect to Home Page.
                return RedirectToAction("Index");
            }
            
        }
	}
}

File: Projects/Software Technologies/Exam Preparation 1/PHP Skeleton/src/AppBundle/Controller/TaskController.php
<?php

namespace AppBundle\Controller;

use AppBundle\Entity\Task;
use AppBundle\Form\TaskType;
use Sensio\Bundle\FrameworkExtraBundle\Configuration\Route;
use Symfony\Bundle\FrameworkBundle\Controller\Controller;
use Symfony\Component\HttpFoundation\Request;

class TaskController extends Controller
{
    /**
     * @param Request $request
     * @Route("/", name="index")
     * @return \Symfony\Component\HttpFoundation\Response
     */
    public function index(Request $request)
    {
        //We use the repository for working with the database and managing our data.
        $taskRepository = $this->getDoctrine()->getRepository(Task::class);

        //Here we want to find all tasks from the databse.
        $tasks = $taskRepository->findAll();

        //And then we give them to the index.html.twig view.
        return $this->render(':task:index.html.twig', ['tasks' => $tasks]);
    }

    /**
     * @param Request $request
     * @Route("/create", name="create")
     * @return \Symfony\Component\HttpFoundation\Response
     */
    public function create(Request $request)
    {
        $task = new Task();

        //$form variable, will create a special token for the user
        // and also take the values from the form the user sent us,
        // AND STICK THEM IN THE $task VARIABLE SO WE CAN WORK WITH THEM!!!
        $form = $this->createForm(TaskType::class, $task);

        //With handleRequest() method we process the request
        //before we check what the user sent us was valid.
        $form->handleRequest($request);

        //Now we need to check two things:
        //1. Did the user sent us a form at all (isSubmitted()).
        //2. Did the user sent us a valid form (isValid()).
        if ($form->isSubmitted() && $form->isValid() ) {
            //We check if the title and comments are null
            if ($task->getTitle() == null || $task->getComments() == null) {
                //If they are we redirect the user to Home Page.
                return $this->redirectToRoute('index');
            }
            //Then we get the entity manager from doctrine
            $em = $this->getDoctrine()->getManager();

            ////and using the “persist” function we add our new task in the database.
            $em->persist($task);
            //Finally, we call the “flush” function, which sends the task to our database.
            $em->flush();

            //
            return $this->redirectToRoute('index');
        }

        //After the task is sent to the database, we redirect the view to the index page of our app.
        return $this->render('task/create.html.twig', ['form' => $form->createView()]);
    }

    /**
     * @Route("/delete/{id}", name="delete")
     *
     * @param $id
     * @param Request $request
     * @return \Symfony\Component\HttpFoundation\Response
     */
    public function delete($id, Request $request)
    {
        //We use the repository for working with the database and managing our data.
        $taskRepository = $this->getDoctrine()->getRepository(Task::class);

        //we want to find the id of the task we want to delete.
        $task = $taskRepository->find($id);

        //We check if the $task variable is null
        if ($task == null) {
            //And if this condition is true we redirect the user to Home Page.
            return $this->redirectToRoute('index');
        }

        // We take the values from the form the user sent us,
        // AND STICK THEM IN THE $task VARIABLE SO WE CAN WORK WITH THEM!!!
        $form = $this->createForm(TaskType::class, $task);

        //With handleRequest() method we process the request
        //before we check what the user sent us was valid.
        $form->handleRequest($request);

        //Now we need to check two things:
        //1. Did the user sent us a form at all (isSubmitted()).
        //2. Did the user sent us a valid form (isValid()).
        if ($form->isSubmitted() && $form->isValid() ) {
            //Then we get the entity manager from doctrine
            $em = $this->getDoctrine()->getManager();

            //We delete the task.
            $em->remove($task);
            //Finally, we call the “flush” function, which sends the task to our database.
            $em->flush();

            //After the task is sent to the database, we redirect the view to the index page of our app.
            return $this->redirectToRoute('index');
        }
         return $this->render('task/delete.html.twig',
             ['task' => $task, 'form' => $form->createView()]);
    }
}
